compiler/rustc_mir/src/transform/instcombine.rs

```diff
@@ -33,6 +33,32 @@ impl<'tcx> MirPass<'tcx> for InstCombine {
             // Then carry out those optimizations.
             MutVisitor::visit_body(&mut InstCombineVisitor { optimizations, tcx }, body);
         }
+        remove_zst_assignments(tcx, body);
+    }
+}
+
+/// Assignments to ZST locals can be eliminated
+fn remove_zst_assignments<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+    let def_id = body.source.def_id();
+    let param_env = tcx.param_env(def_id);
+
+    let local_decls = body.local_decls.clone();
+    for bbds in body.basic_blocks_mut() {
+        for stmt in bbds.statements.iter_mut() {
```

Comment:
instead of iterating, you can use the `retain` method, this way we don't leave nops around to be cleaned up later.

Reply:
Thanks, done